### PR TITLE
Add validation loss

### PR DESCRIPTION
**Pull request by rockerBOO**
Add a validation ratio parameter (splits the training dataset into train/validation by the ratio. Defaults to 0.0 or no split)
Add `loss/val` log

Screenshot 2023-10-30 at 11-31-41 Weights   Biases

Screenshot 2023-10-30 at 11-35-44 Weights   Biases

@kohya-ss I want to propose these changes. Before I spent too much time, wanted to get your feedback. I copied the training process code to do in the validation process. This duplicates the code. Would you have feedback on how you'd want this to happen if you accept this code? I am willing to make the changes. Thank you!

See #193 for more discussion

### CODE REVIEW

**Comment by rockerBOO** [comment]
Thank you Kohya :) I thought that it might be problematic to split the dataset but the code ran so I got a little excited. 

Originally I had the following code that allowed me to pass in a validation dataset.

- `args.validation_dataset` dataset path
- `args.validation_metadata_file` metadata.json `{ "filename.jpg": { "caption": "the image caption" } }, ...`

```python
class ValidationDataset(train_util.FineTuningDataset):
   def __init__(self, image_dir, metadata_file):
        validation_subset = train_util.FineTuningSubset(image_dir=image_dir, metadata_file=metadata_file, num_repeats=1, shuffle_caption=False, keep_tokens=0, color_aug=None, flip_aug=None, face_crop_aug_range=None, random_crop=None, caption_dropout_rate=0.0, caption_dropout_every_n_epochs=0.0, caption_tag_dropout_rate=0.0, caption_prefix="", caption_suffix="", token_warmup_min=0, token_warmup_step=0)
       super().__init__(subsets=[validation_subset], batch_size=1, tokenizer=tokenizer, max_token_length=75, resolution=[768,448], enable_bucket=False, min_bucket_reso=256, max_bucket_reso=1024, bucket_reso_steps=64, bucket_no_upscale=False, debug_dataset=args.debug_dataset)

val_dataset = ValidationDataset(args.validation_dataset, args.validation_metadata_file)
val_dataset.make_buckets()
val_dataset.set_seed(args.seed)
```

It's really rough but is this a direction you would think we should go into? Mostly we can start with a separate validation dataset. I am using the `FineTuningDataset` option (which required a metadata file) but the `DreamboothDataset` option seemed easier as we could just pass a image directory. I'm really only reading the surface of these datasets classes at the moment but wanted to get your feedback first.

**Comment by rockerBOO** [comment]
So I was considering your proposal to do the split inside the dataset when we initialize it.  Maybe we'd want to do the split before we __init__ the class? The problem I'm thinking is once we are inside the dataset class we can't properly create 2 different datasets? We could set the dataset type from the outside but we'd still need to split the dataset up before. 

Maybe we could utilize the [Subset](https://pytorch.org/docs/stable/data.html#torch.utils.data.Subset) from PyTorch which has indices that would represent different dataset indices from the parent dataset. We could then iterate through the subsets. 

Only problem with that approach would be the reg images? At least with the random_split function. 

But we could setup our own Subset with our own indices and work around the reg images in the "validation set". Then we can keep everything else the same way. If I am thinking about that correctly. 

Will wait for your considerations. Thank you :)

**Comment by rockerBOO** [comment]
Additionally this may allow us to have the "validation" dataset be a `[[dataset.subset]]` where we don't split it but the user can pass it (maybe something like `is_validation = true`). Then we can create the 2 torch.utils.data.Subsets (train, validation) from that.

**Comment by rockerBOO** [comment]
For example I have the following code. I am showing an augmented code from PyTorch for random split but limiting it to the only the viable indices.

```python
        def get_indices_without_reg(dataset: torch.utils.data.Dataset):
            return [i for i, (key, item) in enumerate(dataset.image_data.items()) if item.is_reg is False]

        from typing import Sequence, Union
        from torch._utils import _accumulate
        import warnings
        from torch.utils.data.dataset import Subset

        # Augmented from https://pytorch.org/docs/stable/_modules/torch/utils/data/dataset.html#random_split
        def random_split(dataset: torch.utils.data.Dataset, lengths: Sequence[Union[int, float]]):
            indices = get_indices_without_reg(dataset)
            random.shuffle(indices)

            subset_lengths = []

            for i, frac in enumerate(lengths):
                if frac < 0 or frac > 1:
                    raise ValueError(f"Fraction at index {i} is not between 0 and 1")
                n_items_in_split = int(math.floor(len(indices) * frac))
                subset_lengths.append(n_items_in_split)

            remainder = len(indices) - sum(subset_lengths)

            for i in range(remainder):
                idx_to_add_at = i % len(subset_lengths)
                subset_lengths[idx_to_add_at] += 1

            lengths = subset_lengths
            for i, length in enumerate(lengths):
                if length == 0:
                    warnings.warn(f"Length of split at index {i} is 0. "
                                  f"This might result in an empty dataset.")

            if sum(lengths) != len(indices):
                raise ValueError("Sum of input lengths does not equal the length of the input dataset!")

            return [Subset(dataset, indices[offset - length: offset]) for offset, length in zip(_accumulate(lengths), lengths)]
```

And thus would work the same way to split it (ignoring the reg images). 

```python
            train_ratio = 1 - args.validation_ratio
            validation_ratio = args.validation_ratio
            train, val = random_split(
                train_dataset_group,
                [train_ratio, validation_ratio]
            )
```

Note that there are different random (using random.shuffle vs randperm from pytorch). This remains consistent though using a seed (since random.seed is being set). 

Example indices:

```
print(train.indices)
print(val.indices)

[45, 8, 35, 36, 51, 34, 16, 24, 6, 14, 48, 9, 30, 27, 19, 0, 2, 32, 18, 23, 29, 4, 3, 33, 55, 10, 46, 25, 11, 38, 56, 40, 49, 1, 15, 12, 21, 50, 47, 7, 39, 20, 5, 22, 44, 41] 
[42, 31, 28, 17, 53, 26, 52, 37, 43, 13, 54]
```

**Comment by rockerBOO** [comment]
Screenshot 2023-11-05 at 01-46-48 Weights   Biases

Ok, that seemed to have worked. Only thing was the seed factor. Using an alternate seed for the validation makes it a little tricky. I made it work with a `validation_seed` and without (which could use the training seed instead) but let me know what you think of that.

I also added caching the latents for the validation dataset as well and looked of other options that may need to be updated for both options. 

If this looks good so far then we can possibly look into the duplicate code for the training block (the one I copy and pasted over into the validation section). Probably should unify those 2 so they both run the same code. If you have any thoughts on that part I'm happy to implement that.

Thank you! :)

File: train_network.py [hunk 1]
```diff
@@ -869,14 +892,100 @@ def remove_model(old_ckpt_name):
...
+                    loss = loss * loss_weights
+
+                    if args.min_snr_gamma:
+                        loss = apply_snr_weight(loss, timesteps, noise_scheduler, args.min_snr_gamma)
+                    if args.scale_v_pred_loss_like_noise_pred:
+                        loss = scale_v_prediction_loss_like_noise_prediction(loss, timesteps, noise_scheduler)
+                    if args.v_pred_like_loss:
+                        loss = add_v_prediction_like_loss(loss, timesteps, noise_scheduler, args.v_pred_like_loss)
+
+                    loss = loss.mean()  # 平均なのでbatch_sizeで割る必要なし
```
[the remark this answers is not in the document]
Comment (rockerBOO) [reply train_network.py]: Moved these a common function but the following I had to move them to the device for the validation pass (didn't have them for the training pass. Not quite sure why that is the case but something to consider. 

```python
batch["images"].to(device=accelerator.device, ...)
```
and
```python
loss_weights = batch["loss_weights"].to(accelerator.device)
```

**Comment by rockerBOO** [comment]
Ok I moved them over to `process_batch` and tried not to refactor anything so the function signature is a little long. 

I ran into a bug (I think it's a bug at least) with Wandb to produce the charts. They changed how strict their functionality is for the step metrics. They want you to define a step metric for different steps, otherwise it drops it from being reported. Very frustrating to figure out but hopefully we'll get somewhere with that. I have additional code for defining these metrics. See <https://docs.wandb.ai/guides/track/log/customize-logging-axes> and https://github.com/wandb/wandb/issues/6554 for more information.

This following code would also need to be made to define these metrics. Also associated probably with #792 
```python
    accelerator.init_trackers("kohya-train", config=args)

    if args.log_with == "wandb":
        wandb_tracker = accelerator.get_tracker("wandb").tracker
        wandb_tracker.define_metric("loss_validation_current_step")
        wandb_tracker.define_metric("loss/validation_current", step_metric="loss_validation_current_step")

        wandb_tracker.define_metric("loss_validation_average_step")
        wandb_tracker.define_metric("loss/validation_average", step_metric="loss_validation_average_step")

        wandb_tracker.define_metric("loss_epoch_average_step")
        wandb_tracker.define_metric("loss/epoch_average", step_metric="loss_epoch_average_step")
```

**Note for the following charts I had to remove the `, step=x` part of accelerate.log** so they are a little disjointed but showcasing that it does report. I set the code to be the accurate representation that should work once the wandb bug is fixed. Please try it with tensorboard or other trackers to confirm. 

Screenshot 2023-11-05 at 16-35-33 Weights   Biases

**Comment by rockerBOO** [comment]
No worries about reviewing this. Would just want to not deal with too many conflicts so we are good :). I am continuing to use this branch and testing it. Will try to fix little bugs as I find them. 

- If a dataset directory is empty in the training images dataset it will say it's empty. This happens if all the images in the dataset get moved over to the validation dataset. 

> found directory /path/to/dir contains 0 image files
ignore subset with image_dir='/path/to/dir ': no images found / 画像が見つからないためサブセットを無視します

**Comment by rockerBOO** [comment]
Validation of dreambooth method with repeats in respect to regularization images. We separate out the reg images between the 2 datasets (training and validation). This means the validation dataset will have a certain amount of the reg images to process. Is this idea to have the reg images in the validation dataset?

Secondly, we probably want validation to trigger at a certain number of steps. Also or separately with each epoch. During the dreambooth method where you have a lot of repeats of the training dataset, we probably want to be able to do validation during this run. Maybe we use the number of repeats as some indicator of when to do the validation runs or just plain steps.

**Comment by zhchaoxing** [comment]
Hi @rockerBOO , thank you so much for making the validation loss for DB! I need to run finetuning with validation loss, could you please support finetuning as well? Thanks in advance.

**Comment by rockerBOO** [comment]
> Hi @rockerBOO , thank you so much for making the validation loss for DB! I need to run finetuning with validation loss, could you please support finetuning as well? Thanks in advance.

I would love to add it to all the options. I don't have a good way of testing finetuning option or dreambooth so would need help making sure that is working properly if I made the changes. I think initially getting this through and then replicating it for TI and then the other options.

**Comment by markojak** [comment]
Will this be merged at some point?

**Comment by markojak** [comment]
@rockerBOO if I can help you test let me know can run any of the finetuning options locally or in the cloud

**Comment by markojak** [comment]
Any updates?

**Comment by mostlyhuman** [comment]
Was this ever implemented?

**Comment by rockerBOO** [comment]
This was taken a little further in <https://github.com/kohya-ss/sd-scripts/pull/1165> . Haven't tried it though.

I wanted to wait to merge until we got a dev/sd3/flux release out so it could get the proper attention.
